src/syntax/toEdgeQL.ts

```diff
@@ -388,45 +307,287 @@ export function $toEdgeQL(this: any) {
+      }
 
-    for (const child of ctx.withVars.get(withVar)!.childExprs) {
-      if (vars.has(child)) {
-        visit(child);
+      if (expr.__kind__ === ExpressionKind.Select) {
+        if (
+          isObjectType(expr.__element__) &&
+          // Don't walk shape if select is just wrapping an object type expr
+          // that has it's own shape
+          expr.__element__.__shape__ !==
+            (expr.__expr__ as ObjectTypeSet).__element__.__shape__
+        ) {
+          const walkShape = (shape: object) => {
+            for (let param of Object.values(shape)) {
+              if (param.__kind__ === ExpressionKind.PolyShapeElement) {
+                param = param.__shapeElement__;
+              }
+              if (typeof param === "object") {
+                if (!!(param as any).__kind__) {
+                  childExprs.push(...walkExprTree(param as any, expr, ctx));
+                } else {
+                  walkShape(param);
+                }
+              }
+            }
+          };
+          walkShape(expr.__element__.__shape__ ?? {});
+        }
+      } else {
+        // Update
+        const shape: any = expr.__shape__ ?? {};
+
+        for (const _element of Object.values(shape)) {
+          let element: any = _element;
+          if (!element.__element__) {
+            if (element["+="]) element = element["+="];
+            else if (element["-="]) element = element["-="];
+          }
+          childExprs.push(...walkExprTree(element as any, expr, ctx));
+        }
       }
+
+      childExprs.push(...walkExprTree(expr.__expr__, expr, ctx));
+      break;
+    }
+    case ExpressionKind.Delete: {
+      childExprs.push(...walkExprTree(expr.__expr__, parentScope, ctx));
+      break;
     }
+    case ExpressionKind.Insert: {
+      const shape: any = expr.__shape__ ?? {};
 
-    visiting.delete(withVar);
-    unvisited.delete(withVar);
+      for (const element of Object.values(shape)) {
+        childExprs.push(...walkExprTree(element as any, expr, ctx));
+      }
 
-    sorted.push(withVar);
+      childExprs.push(...walkExprTree(expr.__expr__, expr, ctx));
+      break;
+    }
+    case ExpressionKind.InsertUnlessConflict: {
+      if (expr.__conflict__.on) {
+        childExprs.push(...walkExprTree(expr.__conflict__.on, expr, ctx));
+      }
+      if (expr.__conflict__.else) {
+        childExprs.push(...walkExprTree(expr.__conflict__.else, expr, ctx));
+      }
+
+      childExprs.push(...walkExprTree(expr.__expr__, expr, ctx));
+      break;
+    }
+    case ExpressionKind.Group: {
+      const groupingSet = expr.__grouping__ as any as GroupingSet;
+      for (const [_k, groupExpr] of groupingSet.__exprs__) {
+        // this prevents recurring grouping elements from being walked twice
+        // this way, these won't get pulled into with blocks,
+        // which is good because they need to be rendered in `using`
+        const seen: Set<any> = new Set();
```

Comment:
I think we can bypass the whole repeated variable extraction here, since (if I'm understanding it correctly) the `using` block essentially works the same as `with` but actually within the scope of the `group` expression.